File: code/boj/24479.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N,M,R;
int cnt=1;
vector<vector<int>> maps;
vector<int> visit;
vector<int> ans;

void DFS(int v){
  ans[v] = cnt++;
  for(int i=0;i<maps[v].size();i++){
    int cur = maps[v][i];
    if(visit[cur]) continue;
    visit[cur] = 1;
    DFS(cur);
  }
}

int main(void){
  ios::sync_with_stdio(0);
  cout.tie(0);
  cin >> N >> M >> R;
  maps.resize(N+1, vector<int>());
  visit.resize(N+1, 0);
  ans.resize(N+1, 0);
  for(int i=0;i<M;i++){
    int start,end;
    cin >> start >> end;
    maps[start].push_back(end);
    maps[end].push_back(start);
  }
  for(int i=0;i<maps.size();i++){
    sort(maps[i].begin(), maps[i].end());
    // for(int j=0;j<maps[i].size();j++){
    //   cout << maps[i][j] << " ";
    // }
    // cout << "\n";
  }
  visit[R] = cnt;
  DFS(R);
  for(int i=1;i<ans.size();i++)
    cout << ans[i] << "\n";
}

File: code/boj/24479.js
const filePath =
  process.platform === "linux"
    ? "/dev/stdin"
    : "/Users/jeonsiwon/Desktop/coding-interview/code/boj/input.txt";
const input = require("fs")
  .readFileSync(filePath, "utf-8")
  .toString()
  .trim()
  .split("\n");
const [N, M, R] = input.shift().split(" ").map(Number);
const maps = Array.from({ length: N + 1 }, () => []);
const visit = Array.from({ length: N + 1 }, () => 0);
const ans = Array.from({ length: N + 1 }, () => 0);
let cnt = 1;
visit[R] = cnt;
for (let i = 0; i < M; i++) {
  const [start, end] = input[i].split(" ").map(Number);
  maps[start].push(end);
  maps[end].push(start);
}
maps.map((v) => v.sort((a, b) => a - b));

function DFS(v) {
  ans[v] = cnt++;
  const connected = maps[v];
  for (let i = 0; i < connected.length; i++) {
    const cur = connected[i];
    if (visit[cur]) continue;
    visit[cur] = 1;
    DFS(cur);
  }
}

function solution() {
  DFS(R);
  ans.shift();
  console.log(ans.join("\n"));
}
solution();
// console.log(maps);
